src/main/java/com/aws/greengrass/deployment/KernelUpdateDeploymentTask.java

```diff
@@ -38,49 +42,72 @@ public class KernelUpdateDeploymentTask implements DeploymentTask {
     private final ComponentManager componentManager;
+    private final ExecutorService executorService;
+    private final CompletableFuture<DeploymentResult> deploymentResultCompletableFuture;
 
     /**
      * Constructor for DefaultDeploymentTask.
      *
-     * @param kernel Kernel instance
-     * @param logger Logger instance
-     * @param deployment Deployment instance
+     * @param kernel           Kernel instance
+     * @param logger           Logger instance
+     * @param deployment       Deployment instance
      * @param componentManager ComponentManager instance
+     * @param executorService  ExecutorService instance
      */
     public KernelUpdateDeploymentTask(Kernel kernel, Logger logger, Deployment deployment,
-                                      ComponentManager componentManager) {
+                                      ComponentManager componentManager, ExecutorService executorService) {
         this.kernel = kernel;
         this.deployment = deployment;
         this.logger = logger.dfltKv(DEPLOYMENT_ID_LOG_KEY, deployment.getGreengrassDeploymentId());
         this.componentManager = componentManager;
+        this.executorService = executorService;
+        this.deploymentResultCompletableFuture = new CompletableFuture<>();
     }
 
     @SuppressWarnings({"PMD.AvoidDuplicateLiterals"})
     @Override
     public DeploymentResult call() {
+        executorService.execute(this::waitForServicesToStart);
+        DeploymentResult result;
+        try {
+            result = deploymentResultCompletableFuture.get();
+        } catch (InterruptedException | ExecutionException | CancellationException e) {
+            return null;
+        }
+        componentManager.cleanupStaleVersions();
+        return result;
+
+    }
+
+    private void waitForServicesToStart() {
         Deployment.DeploymentStage stage = deployment.getDeploymentStage();
         try {
             List<GreengrassService> servicesToTrack =
                     kernel.orderedDependencies().stream().filter(GreengrassService::shouldAutoStart)
                             .filter(o -> !kernel.getMain().equals(o)).collect(Collectors.toList());
             long mergeTimestamp = kernel.getConfig().lookup("system", "rootpath").getModtime();
-            logger.atDebug().kv("serviceToTrack", servicesToTrack).kv("mergeTime", mergeTimestamp)
+
+            logger.atInfo().kv("serviceToTrack", servicesToTrack).kv("mergeTime", mergeTimestamp)
                     .log("Nucleus update workflow waiting for services to complete update");
-            DeploymentConfigMerger.waitForServicesToStart(servicesToTrack, mergeTimestamp, kernel);
+            DeploymentConfigMerger.waitForServicesToStart(servicesToTrack, mergeTimestamp, kernel,
+                    deploymentResultCompletableFuture);
+            if (deploymentResultCompletableFuture.isCancelled()) {
+                return;
+            }
 
-            DeploymentResult result = null;
```

Comment:
Nit: You can do this initialization outside of `try block`, update result value like how it was before but now, update that in catch blocks as well. In L-147, complete the future with result.

Reply:
updated